prql-compiler/src/sql/srq/mod.rs

```diff
@@ -0,0 +1,86 @@
+//! SQL Relational Query
+//!
+//! This in an internal intermediate representation between RQ and SQL AST.
+//!
+//! For example, RQ does not have a separate node for DISTINCT, but uses [crate::ast::rq::Take] 1 with
+//! `partition`. In [super::preprocess] module, [crate::ast::rq::Transform] take is wrapped into
+//! [SqlTransform], which does have [SqlTransform::Distinct].
+//!
+//! This module also contains compiler from RQ to SRQ.
+
+mod anchor;
+pub mod ast;
+pub mod context;
+mod gen_query;
+pub mod preprocess;
+
+pub(super) use gen_query::compile_query;
+
+#[cfg(test)]
+mod test {
+    use anyhow::Result;
+
+    use super::ast::SqlQuery;
+    use super::*;
+
+    use crate::sql::{Context, Dialect};
+    use crate::{parser::parse, semantic::resolve};
+
+    fn parse_and_resolve(prql: &str) -> Result<(SqlQuery, Context)> {
+        let query = resolve(parse(prql)?)?;
+
+        compile_query(query, Some(Dialect::Generic))
+    }
+
+    fn count_atomics(prql: &str) -> usize {
+        let (query, _ctx) = parse_and_resolve(prql).unwrap();
+
+        query.ctes.len() + 1
+    }
+
+    #[test]
+    fn test_ctes_of_pipeline() {
+        // One aggregate, take at the end
+        let prql: &str = r###"
+        from employees
+        filter country == "USA"
+        aggregate [sal = average salary]
+        sort sal
+        take 20
+        "###;
+
+        assert_eq!(count_atomics(prql), 1);
```

Comment:
Obv don't change, but FYI I've started putting the whole expression in the `assert`s, since then when it fails, it shows the full input (in particular with `insta`)